src/main/java/TcpServer.java

```diff
@@ -0,0 +1,128 @@
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Optional;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TcpServer implements Closeable {
+  // TODO: make logger instance-specific, prefixing logs w/ config & object info
+  static final Logger logger = Logger.getLogger(TcpServer.class.getName());
+
+  // configuration used to initialize the serverSocket
+  final ServerConfig config;
+
+  // number of clients currently connected to the server socket
+  private final AtomicInteger connectionCount;
+
+  // executor for handling of new connections
+  private final ExecutorService connectionHandlerExecutor;
+
+  // Thread where new connections are accepted from serverSocket
+  private Optional<Thread> acceptThread;
+
+  // socket which manages client connections
+  // Initialized to empty, and set to a Socket object once started.
+  private Optional<ServerSocket> serverSocket;
+
+  TcpServer(ServerConfig config) {
+    this.config = config;
+    this.serverSocket = Optional.empty();
+    this.acceptThread = Optional.empty();
+    this.connectionCount = new AtomicInteger(0);
+    this.connectionHandlerExecutor = Executors.newFixedThreadPool(this.config.maxConnections);
+  }
+
+  /**
+   * @see connectionCount
+    */
+  int getConnectionCount() {
+    return connectionCount.get();
+  }
+
+  /**
+   * Start listening on the specified port.
+   *
+   * Must not be called more than once.
+   *
+   * @throws IOException If the underlying socket couldn't bind successfully.
+   */
+  void start() throws IOException {
+    assert !serverSocket.isPresent();
+    ServerSocket socket = new ServerSocket(this.config.port, this.config.maxConnections);
+    this.serverSocket = Optional.of(socket);
+    Thread acceptThread = new Thread(this::acceptConnections);
+    this.acceptThread = Optional.of(acceptThread);
+    acceptThread.start();
+  }
+
+  // Accept and handle new connections
+  private void acceptConnections() {
+    assert acceptThread.isPresent();
+    assert Thread.currentThread().equals(acceptThread.get());
+    assert serverSocket.isPresent();
+
+    ServerSocket serverSocket = this.serverSocket.get();
+    while (!serverSocket.isClosed()) {
+      try {
+        Socket clientConnection = serverSocket.accept();
+        connectionHandlerExecutor.execute(() -> handleConnection(clientConnection));
+      } catch (IOException e) {
+        logger.log(Level.FINE,
+                   "Exception while accepting new connection",
+                   e);
+      }
+    }
+  }
+
+  // Handle new client connections
+  private void handleConnection(Socket connection) {
```

Comment:
Somebody once told me (🎶) that wanting to test private methods is a sign of a class waiting to be extracted. I have that feeling about this method (and I think you do too based on your PR description). So this method definitely feels like a good candidate to move somewhere else, and I think the extraction would be pretty straightforward and would provide a good seam for a fake `Socket`. Not to mention the fact that I expect this is the place this project will grow - since it handles the behavior of the server once a client is connected.

Reply:
Seems we're [on the same page](https://github.com/bgerstle/fabrial/pull/8#discussion_r213838302). I have some ideas on how that could look. Do you think moving it is in the scope of this PR? I'd prefer to factor it out as part of implementing the echo functionality. Without getting too detailed: I envision a `ConnectionHandler` interface, and having an `EchoHandler` which writes data back to clients. Of course, this will eventually be registered using some routing API... etc. etc.